src/json/from_json.cr

```diff
@@ -226,42 +226,55 @@ end
 def Union.new(pull : JSON::PullParser)
   location = pull.location
 
-  # Optimization: use fast path for primitive types
   {% begin %}
-    # Here we store types that are not primitive types
-    {% non_primitives = [] of Nil %}
-
-    {% for type, index in T %}
-      {% if type == Nil %}
-        return pull.read_null if pull.kind.null?
-      {% elsif type == Bool ||
-                 type == Int8 || type == Int16 || type == Int32 || type == Int64 ||
-                 type == UInt8 || type == UInt16 || type == UInt32 || type == UInt64 ||
-                 type == Float32 || type == Float64 ||
-                 type == String %}
-        value = pull.read?({{type}})
-        return value unless value.nil?
-      {% else %}
-        {% non_primitives << type %}
-      {% end %}
+    case pull.kind
+    {% if T.includes? Nil %}
+    when .null?
+      return pull.read_null
+    {% end %}
+    {% if T.includes? Bool %}
+    when .bool?
+      return pull.read_bool
+    {% end %}
+    {% if T.includes? String %}
+    when .string?
+      return pull.read_string
+    {% end %}
+    when .int?
+    {% type_order = [Int64, UInt64, Int32, UInt32, Int16, UInt16, Int8, UInt8, Float64, Float32] %}
+    {% for type in type_order.select { |t| T.includes? t } %}
+      value = pull.read?({{type}})
+      return value unless value.nil?
     {% end %}
+    when .float?
+    {% type_order = [Float64, Float32] %}
+    {% for type in type_order.select { |t| T.includes? t } %}
+      value = pull.read?({{type}})
+      return value unless value.nil?
+    {% end %}
```

Comment:
Do we need this *and* `Float64, Float32` in the above branch??

Reply:
Yes. The `int?` case handles extracting a value if the parser picks up the original value as an integer, but the only numeric in the Union is a float. This then captures float → float.

Reply:
Oh right, I forgot about the `when .float?` because of the indentation.